domain/src/main/java/org/oppia/domain/topic/TopicController.kt

```diff
@@ -191,6 +239,149 @@ class TopicController @Inject constructor(
+    ) {
+      val ongoingTopicList = createOngoingTopicListFromProgress(it)
+      AsyncResult.success(ongoingTopicList)
+    }
+
+    return dataProviders.convertToLiveData(ongoingTopicListDataProvider)
+  }
+
+  fun retrieveQuestionsForSkillIds(skillIdsList: List<String>): DataProvider<List<Question>> {
+    return dataProviders.createInMemoryDataProvider(QUESTION_DATA_PROVIDER_ID) {
+      loadQuestionsForSkillIds(skillIdsList)
+    }
+  }
+
+  private fun createOngoingTopicListFromProgress(
+    topicProgressList: TopicProgressList
+  ): OngoingTopicList {
+    val ongoingTopicListBuilder = OngoingTopicList.newBuilder()
+    topicProgressList.topicProgressList.forEach { topicProgress ->
+      val topic = retrieveTopic(topicProgress.topicId)
+      topic.storyList.forEach { storySummary ->
+        if (topicProgress!!.storyProgressMap.containsKey(storySummary.storyId)) {
+          val storyProgress = topicProgress.storyProgressMap[storySummary.storyId]
+          val lastChapterSummary = storySummary.chapterList.last()
+          if (!storyProgress!!.chapterProgressMap.containsKey(lastChapterSummary.explorationId)) {
+            ongoingTopicListBuilder.addTopic(topic)
+          }
+        }
+      }
+    }
+    return ongoingTopicListBuilder.build()
+  }
+
+  private fun createCompletedStoryListFromProgress(storyProgressList: StoryProgressList): CompletedStoryList {
+    val completedStoryListBuilder = CompletedStoryList.newBuilder()
+    storyProgressList.storyProgressList.forEach { storyProgress ->
+      val storySummary = retrieveStory(storyProgress.storyId)
+      val lastChapterSummary = storySummary.chapterList.last()
+      if (storyProgress.chapterProgressMap.containsKey(lastChapterSummary.explorationId)
+        && storyProgress.chapterProgressMap[lastChapterSummary.explorationId] == ChapterPlayState.COMPLETED
+      ) {
+        completedStoryListBuilder.addStorySummary(storySummary)
+      }
+    }
+    return completedStoryListBuilder.build()
+  }
+
+  /** Combines the specified topic without progress and topic-progress into a topic. */
+  private fun combineTopicAndTopicProgress(topic: Topic, topicProgress: TopicProgress): Topic {
+    val topicBuilder = topic.toBuilder()
+    if (topicProgress.storyProgressMap.isNotEmpty()) {
+      topic.storyList.forEachIndexed { storyIndex, storySummary ->
+        val updatedStorySummary = if (topicProgress.storyProgressMap.containsKey(storySummary.storyId)) {
+          combineStorySummaryAndStoryProgress(storySummary, topicProgress.storyProgressMap[storySummary.storyId]!!)
+        } else {
+          setFirstChapterAsNotStarted(storySummary)
+        }
+        topicBuilder.setStory(storyIndex, updatedStorySummary)
+      }
+    } else {
+      topic.storyList.forEachIndexed { storyIndex, storySummary ->
+        val updatedStorySummary = setFirstChapterAsNotStarted(storySummary)
+        topicBuilder.setStory(storyIndex, updatedStorySummary)
```

Comment:
Similar for here and elsewhere: I suggest making this functional code more side effect free.

Reply:
@BenHenning I think this I cannot change by that similar code that I had written earlier because the `topicBuilder.setStory` needs an index and therefore each and every item will need to be laid out by iterating.